text/0000-sbom.md

```diff
@@ -0,0 +1,125 @@
+
+Quoting [NTIA Whitepaper][NTIA-WP] - 
+> Modern software systems involve increasingly complex and dynamic supply chains. Unfortunately, the composition and functionality of these systems lacks transparency; this contributes substantially to cybersecurity risks, alongside the cost of development, procurement, and maintenance. This has broad implications in our interconnected world; risk and cost affect collective goods, like public safety and national security, in addition to the products and services upon which businesses rely.
+> We propose that increased supply chain transparency through SBOMs can reduce cybersecurity risks and overall costs by:
+> - Enhancing the identification of vulnerable systems and the root cause of incidents
+> - Reducing unplanned and unproductive work
+> - Supporting more informed market differentiation and component selection
+> - Reducing duplication of effort by standardizing formats across multiple sectors
+> - Identifying suspicious or counterfeit software components
+
+Currently since container images built by Cloud Native Buildpacks don't follow a generic file system layout, they may not be easily scannable by container scanning tools. We do however provide a BOM which should greatly help sidestep the entire manual scanning process and speed up things like CVE detection by directly providing the CVE scanner with a BOM.
+
+However, since we currently do not impose any standard for what the BOM should look like and since the metadata table in BOM is a freeform table, it is very hard to have consistent BOMs that can be used by CVE scanners.
+
+Having a structured and consistent SBOM that can be consumed by downstream systems could greatly enhance the security proposition around buildpacks.
+
+## What use cases does it support?
+
+- Consistent SBOM across buildpacks
+- Improved integration with other supply chain analysis and scanning tools
+
+## What is the expected outcome?
+
+- Use of CycloneDX as the SBOM storage format and updates the the specification and lifecycle to reflect the same.
+- Moving the SBOM from `<layer>.toml`, `launch.toml` and `build.toml` to a separate file.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Currently the SBOM is defined in the `<layer>.toml`, `launch.toml` and `build.toml` respectively under the `[bom]` table.
+It may not be the most user-friendly way for buildpack authors to create SBOM documents in the above format. This RFC proposes that the `bom` be moved to `<layer>.bom.<ext>`, `launch.bom.<ext>` and `build.bom.<ext>` instead where `<ext>` will be `cdx.json`
+
+The above bom documents need to be created in CycloneDX format.
+
+The lifecycle of the respective `bom` files during rebuilds would be exactly the same as the `bom` table currently. 
+
+The lifecycle would be responsible for taking all of the above `bom` files, merging it while tagging it with the originating layer or buildpack the `bom` came from and storing it in `/layers/config/bom.cdx.json`.
```

Comment:
I am interested in the fact that all of the BOM files are merged into seemingly one BOM file. How do I distinguish things that were in the `build` BOM from those in the `launch` BOM? Is there a distinction at all in the final merged BOM? If there is no distinction why bother separating them above?

Reply:
There is a distinction. Each bom component entry has the following metadata properties added to it. https://github.com/samj1912/rfcs/blob/sbom/text/0000-sbom.md#how-it-works

Reply:
> io.buildpacks.bom.buildpack.id - Buildpack ID for the buildpack that created the BOM
> io.buildpacks.bom.types - Set to a comma seperated list of build or launch as applicable.
> io.buildpacks.bom.layer.name (Optional) - Set to the name of the layer if the bom was associated with a specific layer.

Reply:
Given the recent conversation in the OH - I guess we want these two to be separate files instead of just being able to distinguish between the two. Will make that change.

Reply:
How we will handle SBOM merging when a build contains both older and newer buildpacks that are producing freeform toml and cyclone dx BOMs respectively?

I can't think of many elegant ways to do this so we may need to do something dramatic? e.g. maybe one of
- support two different "merged" boms if we have two types of data
- have a hard cutoff in the platform API. If platform api is 0.x or above we produce a cycloneDX bom with whatever data we have in that format. If it is below we produce a CNB BOM with the same limitations. If we do this we may want to consider allowing newer buildpacks to produce data in both formats for a time to retain functionality on older platforms.

I kinda hate both of those options. Does anyone have a better idea? If not it may be a necessary evil.

Reply:
@ekcasey That's a good point I hadn't considered. I don't HATE option 2 that you've proposed, but obviously I'd love to find a state in which we don't have to support both formats. 

I think the idea of supporting two different merged BOMs in bullet # 1 would be super confusing for someone to try to look at. It would also just be kind of hard to use scanning tools if the BOMs came in two formats but were both needed to make up the entire set of metadata.

Reply:
We could convert the existing free-form TOML to CycloneDX by placing the arbitrary TOML in the `properties` field of a `component`.